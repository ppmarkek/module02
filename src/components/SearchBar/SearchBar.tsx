import React, { useEffect, useState } from 'react';
import { ComponentsReducer } from '../../types';
import { getAllRequestedResults } from '../../requests';
import { useSearchParams } from 'react-router-dom';
import './styles.css';
import { useDispatch, useSelector } from 'react-redux';
import {
  setLoading,
  setResults,
  setSearchResult,
} from '../../redux/componentsReducer';

export const SearchBar = () => {
  const dispatch = useDispatch();
  const savedCriteria = useSelector(
    (state: ComponentsReducer) => state.componentsReducer.searchResult
  );

  const [searchParams, setSearchParams] = useSearchParams();

  const [inputValue, setInputValue] = useState(savedCriteria ?? '');
  const [itemsPerPage, setItemsPerPage] = useState('10');

  const currentPageQuery = searchParams.get('page');
  const currentSizeQuery = searchParams.get('size');

  const handleSearch = () => {
    dispatch(setLoading(true));
    dispatch(setSearchResult(inputValue));
    setSearchParams({
      ...searchParams,
      page: (0).toString(),
      size: parseInt(itemsPerPage),
    });
    getAllRequestedResults({
      searchCriteria: savedCriteria?.length ? savedCriteria : inputValue,
      size: currentSizeQuery ?? itemsPerPage,
      page: currentPageQuery ?? '1',
    }).then((response) => {
      dispatch(setResults(response));
      dispatch(setLoading(false));
    });
  };

  const throwError = () => {
    throw new Error('Intentional error for testing ErrorBoundary');
  };

  useEffect(() => {
    dispatch(setLoading(true));
    getAllRequestedResults({
      searchCriteria: inputValue,
      size: currentSizeQuery ?? itemsPerPage,
      page: currentPageQuery ?? '1',
    }).then((response) => {
      dispatch(setResults(response));
      dispatch(setLoading(false));
    });
  }, [currentPageQuery, itemsPerPage]);

  return (
    <div className="searchInputsWrapper">
      <div className="searchBar" data-testid="search-bar">
        <label htmlFor="searchInput" id="SearchTitle">
          Search:
        </label>
        <input
          name="searchInput"
          className="searchInput"
          defaultValue={inputValue}
          onChange={(window) => setInputValue(window.target.value)}
        />
        <button className="searchButton" onClick={() => handleSearch()}>
          show
        </button>
        <button onClick={() => throwError()} id="errorButton">
          Error
        </button>
      </div>
      <div className="searchSizer" data-testid="searchSizer">
        <label htmlFor="searchSizeInput" id="SearchTitle">
          Items per page:
        </label>
        <input
          name="searchSizeInput"
          defaultValue={itemsPerPage}
          onChange={(window) => setItemsPerPage(window.target.value)}
        />
        <button onClick={() => handleSearch()}>show</button>
      </div>
    </div>
  );
};
