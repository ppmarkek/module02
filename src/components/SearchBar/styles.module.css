.searchBar {
  height: 5%;
  width: auto;
  display: flex;
  align-items: center;
}

.SearchTitle {
  padding-right: 10px;
}

.searchBar input {
  all: unset;
  padding: 0px 10px;
  border-radius: 15px 0px 0px 15px;
  height: 40px;
  width: 500px;
  border: 1px solid #8083a3;
  transition: all 0.5s;
}

.searchBar input:hover {
  border: 1px solid #6b59cc;
}

.searchBar input:focus {
  border: 1px solid #6b59cc;
}

.searchButton {
  height: 42px;
  cursor: pointer;
  border: 1px solid #8083a3;
  border-left: unset;
  background: #fff;
  color: #000;
  transition: all 0.5s;
}

.searchButton:hover {
  background: #6b59cc;
  color: #fff;
}

.searchSizer {
  height: 50px;
  display: flex;
  align-items: center;
  padding-left: 20px;
}

.searchSizer button {
  height: 42px;
  cursor: pointer;
  border: 1px solid #8083a3;
  border-left: unset;
  border-radius: 0px 15px 15px 0px;
  background: #fff;
  color: #000;
  transition: all 0.5s;
}

.searchSizer button:hover {
  background: #6b59cc;
  color: #fff;
}

.searchSizer input {
  all: unset;
  padding: 0px 10px;
  border-radius: 15px 0px 0px 15px;
  height: 40px;
  width: 80px;
  border: 1px solid #8083a3;
  transition: all 0.5s;
}

.searchInputsWrapper {
  display: flex;
  align-items: center;
  justify-content: center;
}

.errorButton {
  height: 42px;
  cursor: pointer;
  border-radius: 0px 15px 15px 0px;
  border: 1px solid #e60606;
  border-left: unset;
  background: #fff;
  color: #000;
  transition: all 0.5s;
}
.errorButton:hover {
  background: #e60606;
  color: #fff;
}
