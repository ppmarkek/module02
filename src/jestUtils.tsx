import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { reducer } from './redux/index';
import { createStore } from '@reduxjs/toolkit';

export const renderWithRedux = (ui: React.JSX.Element, state = {}) => {
  const Wrapper = ({
    children,
  }: {
    children: React.ReactNode | React.ReactNode[];
  }) => (
    <Router>
      <Provider store={createStore(reducer, state)}>{children}</Provider>
    </Router>
  );

  return render(ui, { wrapper: Wrapper });
};
