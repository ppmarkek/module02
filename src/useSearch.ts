import { useSelector } from 'react-redux';
import { AppState } from './types';

export const useSearch = () => {
  const requestResults = useSelector(
    (state: AppState) => state.componentsReducer.results
  );
  const isLoading = useSelector(
    (state: AppState) => state.componentsReducer.isLoading
  );
  const details = useSelector(
    (state: AppState) => state.componentsReducer.details
  );
  const searchResult = useSelector(
    (state: AppState) => state.componentsReducer.searchResult
  );

  const isDetailsLoading = useSelector(
    (state: AppState) => state.componentsReducer.isDetailsLoading
  );

  const isShowDetails = useSelector(
    (state: AppState) => state.componentsReducer.isShowDetails
  );

  const isError = useSelector(
    (state: AppState) => state.componentsReducer.isError
  );

  return {
    requestResults,
    isLoading,
    isDetailsLoading,
    details,
    searchResult,
    isShowDetails,
    isError,
  };
};
