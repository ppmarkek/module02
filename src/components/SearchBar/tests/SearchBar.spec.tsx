import { renderWithRedux } from '../../../jestUtils';
import { appStoreMock, useSearchHookMock } from '../../../fixtures/results';
import * as useSearch from '../../../useSearch';
import '@testing-library/jest-dom';
import * as router from 'react-router';
import { fireEvent, waitFor } from '@testing-library/react';
import { SearchBar } from '../SearchBar';

const navigate = jest.fn();

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('../../../redux/apiService', () => {
  const mockData = {
    results: [
      { id: 1, title: 'Test Item 1' },
      { id: 2, title: 'Test Item 2' },
    ],
  };

  const mockCategoriesData = {
    categories: ['Category 1', 'Category 2', 'Category 3'],
  };
  const originalModule = jest.requireActual('../../../redux/apiService');

  const mockUseGetAllRequestedResultsQuery = jest.fn().mockReturnValue({
    data: mockData,
    refetch: jest.fn().mockResolvedValue({ data: mockData }),
  });

  const mockUseGetCategoriesQuery = jest.fn().mockReturnValue({
    data: mockCategoriesData,
    refetch: jest.fn().mockResolvedValue({ data: mockCategoriesData }),
  });

  return {
    ...originalModule,
    useGetAllRequestedResultsQuery: mockUseGetAllRequestedResultsQuery,
    useGetCategoriesQuery: mockUseGetCategoriesQuery,
  };
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: jest
    .fn()
    .mockReturnValue([new URLSearchParams(), jest.fn()]),
}));

describe('SearchBar component', () => {
  it('should render correctly', () => {
    jest
      .spyOn(useSearch, 'useSearch')
      .mockImplementation(() => useSearchHookMock);

    const component = renderWithRedux(<SearchBar />, {
      appStore: appStoreMock,
    });

    expect(component).toMatchSnapshot();
  });

  it('should show search results and navigate with query on search button', () => {
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
    jest
      .spyOn(useSearch, 'useSearch')
      .mockImplementation(() => useSearchHookMock);

    const { getByTestId } = renderWithRedux(<SearchBar />);
    fireEvent.click(getByTestId('search-button'));
  });

  it('should show search results and navigate with query on show per page button', async () => {
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
    jest
      .spyOn(useSearch, 'useSearch')
      .mockImplementation(() => useSearchHookMock);

    const { getByTestId } = renderWithRedux(<SearchBar />);
    const showPerPageButton = getByTestId('show-per-page-button');

    fireEvent.click(showPerPageButton);

    await waitFor(() => {
      expect(navigate).toHaveBeenCalledWith('?page=0&size=10', undefined);
    });
  });

  it('should handle error correctly on error test button', () => {
    const { getByTestId } = renderWithRedux(<SearchBar />);
    fireEvent.click(getByTestId('show-error-button'));

    expect(() => {
      fireEvent.click(getByTestId('show-error-button'));
    }).toThrow('Intentional error for testing ErrorBoundary');
  });

  it('should update input value on change', () => {
    const { getByTestId } = renderWithRedux(<SearchBar />);
    const searchInput = getByTestId('search-input') as HTMLInputElement;

    fireEvent.change(searchInput, { target: { value: 'new search' } });
    expect(searchInput.value).toBe('new search');
  });
});
