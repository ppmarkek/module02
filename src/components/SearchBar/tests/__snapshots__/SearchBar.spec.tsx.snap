// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="searchInputsWrapper"
      >
        <div
          class="searchBar"
          data-testid="search-bar"
        >
          <label
            for="searchInput"
            id="SearchTitle"
          >
            Search:
          </label>
          <input
            class="searchInput"
            data-testid="search-input"
            name="searchInput"
            value="people"
          />
          <button
            class="searchButton"
            data-testid="search-button"
          >
            Show
          </button>
          <button
            data-testid="show-error-button"
          >
            Test Error
          </button>
        </div>
        <div
          class="searchSizer"
          data-testid="searchSizer"
        >
          <label
            for="searchSizeInput"
            id="SearchTitle"
          >
            Items per page:
          </label>
          <input
            class="searchSizeInput"
            name="searchSizeInput"
            value="10"
          />
          <button
            data-testid="show-per-page-button"
          >
            Show
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="searchInputsWrapper"
    >
      <div
        class="searchBar"
        data-testid="search-bar"
      >
        <label
          for="searchInput"
          id="SearchTitle"
        >
          Search:
        </label>
        <input
          class="searchInput"
          data-testid="search-input"
          name="searchInput"
          value="people"
        />
        <button
          class="searchButton"
          data-testid="search-button"
        >
          Show
        </button>
        <button
          data-testid="show-error-button"
        >
          Test Error
        </button>
      </div>
      <div
        class="searchSizer"
        data-testid="searchSizer"
      >
        <label
          for="searchSizeInput"
          id="SearchTitle"
        >
          Items per page:
        </label>
        <input
          class="searchSizeInput"
          name="searchSizeInput"
          value="10"
        />
        <button
          data-testid="show-per-page-button"
        >
          Show
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
